src/apps/frontend/contexts/async.hook.ts

```diff
@@ -0,0 +1,55 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import {
+  UseAsyncResponse,
+  AsyncResult,
+  AsyncError,
+} from '../types';
+
+const useAsync = <T>(
+  asyncFn: (...args: any[]) => Promise<AsyncResult<T>>,
+  immediate = false,
+): UseAsyncResponse<T> => {
+  const [result, setResult] = useState<T | undefined>(undefined);
+  const [isLoading, setLoading] = useState(false);
+  const [error, setError] = useState<AsyncError | undefined>(undefined);
+
+  const asyncCallback = useCallback(
+    async (...args) => {
+      setError(undefined);
+      setLoading(true);
+      try {
+        const response = await asyncFn(...args);
+
+        if (response.data) {
+          setResult({ ...response.data });
+        }
+
+        return response?.data;
+      } catch (e) {
+        const err = new AsyncError(e?.response?.data);
+        setError(err);
+        throw new Error(err.message);
+      } finally {
+        setLoading(false);
+      }
+    },
+    [asyncFn],
+  );
+
+  useEffect(() => {
+    if (immediate) {
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      asyncCallback();
+    }
+  }, [asyncCallback, immediate]);
```

Comment:
Why is this needed? I see `immediate` as always false

Reply:
If we want the function to run when the component mounts, we can set this flag to true. That way, we don't have to call the asyncCallback ourselves.

Reply:
For instance, if we need to retrieve the profile information of the logged-in user, we can directly use:

`const {isLoading, error, result } = useAuthContext(userProfile)`

There's no requirement to write it like this:

```
const {isLoading, error, result, userProfileCallback } = useAuthContext(userProfile)

userProfileCallback()
```

Reply:
Not using it at the moment.